presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -261,13 +261,15 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
                     .collect(toImmutableList());
 
             List<List<Symbol>> inputs = new ArrayList<>();
+            Map<List<Symbol>, Symbol> inputsToOutputs = new HashMap<>();
```

Comment:
It is much better now. It is acceptable to me, but I would do it differently ;)

I would do two `ExchangeNode` transformations. First I would create `ExchangeNode` with canonical inputs and outputs (the old `visitExchangeNode`). Then I would remove input duplicates.